File: algorithms/2_medium/S/01081_smallest_subsequence_of_distinct_characters/stack.c
char* smallestSubsequence(const char* s)
{
    const size_t len = strlen(s);
    assert(len >= 1 && len <= 1000);
    char* stack = malloc((len + 1) * sizeof(*s));
    size_t stack_len = 0;
    assert(stack != NULL);
    uint16_t freqs[26] = { 0 };

    for (size_t i = 0; i < len; i += 1) {
        const char ch = s[i];
        assert(ch >= 'a' && ch <= 'z');
        freqs[ch - 'a'] += 1;
    }

    bool seen[26] = { false };

    for (size_t i = 0; i < len; i += 1) {
        const char ch = s[i];
        if (!seen[ch - 'a']) {
            while (stack_len > 0 && stack[stack_len - 1] >= ch && freqs[stack[stack_len - 1] - 'a'] >= 1) {
                seen[stack[stack_len - 1] - 'a'] = false;
                stack_len -= 1;
            }
            seen[ch - 'a'] = true;
            stack[stack_len] = ch;
            stack_len += 1;
        }
        freqs[ch - 'a'] -= 1;
    }

    stack[stack_len] = '\0';
    return stack;
}

File: algorithms/2_medium/S/01081_smallest_subsequence_of_distinct_characters/stack.rs
impl Solution {
    pub fn smallest_subsequence(s: String) -> String {
        let bytes = s.as_bytes();
        let len = bytes.len();
        assert!(len >= 1 && len <= 1000);
        let mut stack: Vec<u8> = Vec::with_capacity(len);
        let mut freqs = [0_u16; 26];
        for &b in bytes {
            assert!(b >= b'a' && b <= b'z');
            freqs[(b - b'a') as usize] += 1;
        }
        let mut seen = [false; 26];
        for &b in bytes {
            if !seen[(b - b'a') as usize] {
                while let Some(top) = stack.last() && *top >= b && freqs[(*top - b'a') as usize] >= 1 {
                    seen[(*top - b'a') as usize] = false;
                    stack.pop();
                }
                seen[(b - b'a') as usize] = true;
                stack.push(b);
            }
            freqs[(b - b'a') as usize] -= 1;
        }
        String::from_utf8(stack).unwrap()
    }
}
